### PR TITLE
XTypes: Integrate complete  TypeObject with builtin topics

### PR DESCRIPTION
**Pull request by sonndinh**

### CODE REVIEW

**Review by ClaytonCalabrese — CHANGES_REQUESTED** [review]
There are a bunch of test failures in the CI. I'd recommend merging from master and then trying to figure out what is causing the failures.

Seems like a bunch of compiler tests as well as the XTypes test are experiencing problems.

File: dds/DCPS/DiscoveryBase.h [hunk 1]
```diff
@@ -1247,40 +1271,56 @@ namespace OpenDDS {
...
-        md.writer = writer;
-        md.reader = reader;
+        // If the type object is not in cache, send RPC request
         md.time_added_to_map = MonotonicTimePoint::now();
 
+        // NOTE(sonndinh): Is it possible for a discovered endpoint to include only the "complete"
+        // part in its TypeInformation? If it's possible, then we may need to handle that case, i.e.,
+        // request only the remote complete TypeObject (if it's not already in the cache).
+        // The following code assumes when the "minimal" part is not included in the discovered
+        // endpoint's TypeInformation, then the "complete" part also is not included.
```
[the remark this answers is not in the document]
Comment (sonndinh) [reply dds/DCPS/DiscoveryBase.h:1281]: I talked to Adam about this but I think we haven't had a conclusive answer. Our matching process is assuming the `minimal` member is always present in `TypeInformation` but the spec seems not exclude the case when only the `complete` member is present. We probably need a further discussion on this.

File: dds/DCPS/RTPS/Sedp.cpp [hunk 2]
```diff
@@ -4606,6 +4650,15 @@ Sedp::DiscoveryReader::data_received_i(const DCPS::ReceivedDataSample& sample,
...
     }
 #endif
+
+    // Cache non-empty TypeInformation
+    if (wdata.type_info_.minimal.typeid_with_size.type_id.kind() != XTypes::TK_NONE ||
+        wdata.type_info_.complete.typeid_with_size.type_id.kind() != XTypes::TK_NONE) {
+      const GUID_t& remote_guid = wdata.writer_data_.writerProxy.remoteWriterGuid;
+      DDS::BuiltinTopicKey_t key = DCPS::repo_id_to_bit_key(remote_guid);
+      sedp_.type_lookup_service_->cache_type_info(key, wdata.type_info_);
+    }
```
[the remark this answers is not in the document]
Comment (sonndinh) [reply dds/DCPS/RTPS/Sedp.cpp:4660]: I think it's kind of small to put into a function.

File: dds/DCPS/Service_Participant.cpp [hunk 3]
```diff
@@ -2878,6 +2878,36 @@ Service_Participant::bit_autopurge_disposed_samples_delay(const DDS::Duration_t&
...
+{
+  DomainParticipantImpl* participant_servant = dynamic_cast<DomainParticipantImpl*>(participant);
+  if (participant_servant) {
+    XTypes::TypeLookupService_rch tls = participant_servant->get_type_lookup_service();
+    if (tls) {
+      return tls->get_type_objects(ti);
+    }
+  }
+
+  return XTypes::TypeObject();
```
[the remark this answers is not in the document]
Comment (sonndinh) [reply dds/DCPS/Service_Participant.cpp:2908]: Fixed

**Comment by sonndinh** [comment]
GitHub shows incorrect commit history for this PR after I merged master to the xtypes_dynamic branch, and the xtypes_dynamic branch to the PR's branch. So I opened PR #2876 to replace this PR. The new PR shows the correct commit history.

**Pull request closed without merging**